### PR TITLE
Trovi 2.0: Implement read endpoints

### PR DESCRIPTION
**Pull request by super-cooper**
Trovi 2.0: Implement read endpoints

Implements the `ListArtifacts` and `GetArtifact` endpoints. The original task called for auth to be implemented by this point,
but since this task was prioritized first, auth has been left as a TODO. 
There is also a basic test to confirm that the models will properly serialize to JSON, 
but better testing will be implemented while the code is in review. :heavy_check_mark: 

## Changes

 * Implements an API app
 * Creates some basic response types which make error responses a little easier
 * ~~Implements read endpoints as classes with a static `get` function, ensuring that only GET requests are valid~~
 * Fixes an unforeseen bug in `URNField`, which did not handle null values
 * ~~Adds an abstract base `APIModel`, which defines how models will be serialized to JSON.~~
   ~~All models which appear in API responses now implement this class.~~
     * ~~Defines an `api_items` property which defines the structure of how a model will be translated to a JSON API response~~
     * ~~Defines a `to_json` property which defines how specifically the model will be serialized.~~
       ~~The implementation in the base class adheres strictly to the properties of `api_items`~~
      ~~and will be the correct choice in almost every case.~~
 * Adds some JSON Protocol type definitions for use in hints

### NOTE
~~This PR is currently a little weird because it is based off of the work in #3, which has not been merged into main yet. I will pull it out of draft state once that work is on main so that the commits on this branch are not so redundant.~~

Many of the changes above have been made obsolete by the integration with [Django Rest Framework](https://www.django-rest-framework.org). All structures are now implemented via the interfaces provided by the rest framework, and the API behaves to the design spec.

### CODE REVIEW

**Comment by super-cooper** [comment]
[Trovi 2.0: Implement read endpoints](https://trello.com/c/T1NkWm6t/2480-trovi-20-implement-read-endpoints)

**Comment by diurnalist** [comment]
I merged #3, so you can try rebasing against the main branch again to clean up that aspect of the history :)

File: docker/Dockerfile [hunk 1]
```diff
@@ -2,13 +2,12 @@ ARG PY_IMG_TAG=3.9.7
...
 FROM python:${PY_IMG_TAG}
 
-COPY . /project
 WORKDIR /project
 
 VOLUME ["/media"]
 VOLUME ["/static"]
 
-COPY poetry.lock pyproject.toml /setup/
+COPY poetry.lock pyproject.toml /project/
```
Comment (diurnalist) [inline docker/Dockerfile:10]: The code doesn't actually get copied into the image, does it? This will not work for prod when there is no bind-mount. If you want to avoid cache busting, you can do this:

```dockerfile
FROM python:${PY_IMG_TAG} as base

# (existing Dockerfile contents)

FROM base as release

COPY . /project
```

Then you can refer to the "base" target in the docker-compose environment or otherwise have a separate step to build the "full" release image separately.

File: docker/dev.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
+ARG PY_IMG_TAG=3.9.7
```
Comment (diurnalist) [inline docker/dev.Dockerfile:1]: Is there a difference b/w this one and the other one? I can't detect one.

File: trovi/models.py [hunk 3]
```diff
@@ -17,16 +89,20 @@ class Artifact(models.Model):
...
+    short_description = models.CharField(
+        max_length=settings.ARTIFACT_SHORT_DESCRIPTION_MAX_CHARS, blank=True, null=True
+    )
+    long_description = models.TextField(
+        max_length=settings.ARTIFACT_LONG_DESCRIPTION_MAX_CHARS
+    )
 
     # Timestamps
-    created_at = models.DateTimeField(auto_now_add=True)
+    created_at = models.DateTimeField(auto_now_add=True, editable=False)
```
Comment (diurnalist) [inline trovi/models.py:101]: Isn't this already not editable by virtue of having `auto_now_add`?

File: trovi/models.py [hunk 4]
```diff
@@ -49,18 +125,39 @@ class Visibility(models.TextChoices):
         default=Visibility.PRIVATE,
     )
     sharing_key = models.CharField(
-        max_length=32, null=True, default=lambda: secrets.token_urlsafe(nbytes=32)
+        max_length=settings.SHARING_KEY_LENGTH * 2,
```
Comment (diurnalist) [inline trovi/models.py:128]: why x2? If there is a reason, maybe a comment is good here. It's non-intuitive.

File: trovi/models.py [hunk 5]
```diff
@@ -70,8 +167,43 @@ def access_count(self) -> int:
...
+        is an index starting at 1 which increments automatically for each version
+        published on the same given day.
 
-class ArtifactEvent(models.Model):
+        The slug is stored in the instance's slug field and saved
+        """
+        if created:
+            time_stamp = instance.created_at.strftime("%Y-%m-%d")
+            versions_today = ArtifactVersion.objects.filter(
+                artifact__created_at__year=instance.created_at.year,
```
Comment (diurnalist) [inline trovi/models.py:182]: This is simpler, I think equivalent, and reduces from 3 kwargs to one:

```suggestion
                artifact__created_at__date=instance.created_at.date()
```

File: trovi/settings.py [hunk 6]
```diff
@@ -140,12 +144,24 @@
 # Database
 # https://docs.djangoproject.com/en/3.2/ref/settings/#databases
 
-DATABASES = {
-    "default": {
-        "ENGINE": "django.db.backends.sqlite3",
-        "NAME": BASE_DIR / "db.sqlite3",
+if os.environ.get("DB_NAME"):
```
Comment (diurnalist) [inline trovi/settings.py:147]: It feels like `DB_HOST` is more the signal here, DB_NAME could default to `trovi`. Though, one could argue DB_HOST could be assumed to be localhost. Might make more sense to just have `DB_ENGINE` or `DB_BACKEND`, which defaults to `mysql` but can be overridden to `sqlite3`.

File: trovi/urls.py [hunk 7]
```diff
@@ -14,8 +14,9 @@
     2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
 """
 from django.contrib import admin
-from django.urls import path
+from django.urls import path, include
 
 urlpatterns = [
     path("trovi-admin-portal/", admin.site.urls),
+    path("artifacts/", include("trovi.api.urls")),
```
Comment (diurnalist) [inline trovi/urls.py:21]: what about "api/"?

**Review by diurnalist — APPROVED** [review]
Intentionally non-deterministic test cases feel quite icky to me, but I'm willing to see where they go. I can imagine that they can be flaky, but the flakiness is also catching edge cases. This strategy resembles property-based testing, which I think is a stronger approach, but to be honest I'm not sure how to express these tests using such a methodology. We'll see how it goes! The issue with non-deterministic tests is it's hard to replicate a failure. So we'll see CI fail and it may be difficult to reproduce the conditions w/o hacking up the test a bit.

File: trovi/settings.py (hunk 6 above)
Comment (super-cooper) [reply trovi/settings.py]: I think sqlite as the default is better, since it requires basically zero configuration, and might be better for testing, since MariaDB has been giving me difficulty.

File: trovi/urls.py (hunk 7 above)
Comment (super-cooper) [reply trovi/urls.py:21]: In the design dock, the endpoint was "artifacts/". If you want to change it though, I am fine with it.

File: trovi/models.py (hunk 4 above)
Comment (super-cooper) [reply trovi/models.py]: Forgot to reply to this. The sharing key is 32 bytes, but it is encoded in base64, so it ends up longer than that. The base64 length calculation puts it at about 1.3x the length of the data, but to be safe, I went with 2x. Now, I've decided to use the actual length calculation, and I've added a comment.

**Comment by super-cooper** [comment]
Read endpoints are 100% in the state I want them to be in. Might be a good idea to give them another once-over. The write-endpoints PR that is coming soon will be based on this work.

File: docker/dev.Dockerfile (hunk 2 above)
Comment (diurnalist) [reply docker/dev.Dockerfile:1]: I would just do this:

```
FROM release as dev

RUN poetry install --dev-only
```

When we build for prod you can do `--target release` to stop at the release stage; by default if you build locally it will install dev deps.

File: trovi/api/filters.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    def filter_queryset(
+        self, request: Request, queryset: models.QuerySet, view: views.View
+    ) -> models.QuerySet:
+        sort_by = request.query_params.get("sort_by")
+
+        if sort_by is not None:
+            if sort_by == "date":
+                query = queryset.order_by("created_at")
+            elif sort_by == "access_count":
+                query = queryset.annotate(
```
Comment (diurnalist) [inline trovi/api/filters.py:24]: cool :)

File: trovi/api/filters.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+                    access_count=Sum("versions__access_count")
+                ).order_by("-access_count")
+            else:
+                raise ValidationError(
+                    f"ListArtifact: Unknown 'sort_by' key. ({sort_by})"
+                )
+        else:
+            # By default, artifacts are sorted by UUID
+            query = queryset.order_by("uuid")
```
Comment (diurnalist) [inline trovi/api/filters.py:33]: This is a bit unexpected, maybe ordering by date by default makes the most sense?

File: trovi/api/serializers.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    def to_representation(self, instance: Artifact) -> dict:
+        return {
+            "id": instance.uuid,
+            "created_at": instance.created_at.strftime(settings.DATETIME_FORMAT),
+            "updated_at": instance.created_at.strftime(settings.DATETIME_FORMAT),
+            "title": instance.title,
+            "short_description": instance.short_description,
+            "long_description": instance.long_description,
+            "tags": ArtifactTagSerializer(instance.tags.all(), many=True).data,
```
Comment (diurnalist) [inline trovi/api/serializers.py:90]: What does `many` do here?

File: trovi/models.py [hunk 11]
```diff
@@ -49,18 +56,23 @@ class Visibility(models.TextChoices):
         default=Visibility.PRIVATE,
     )
     sharing_key = models.CharField(
-        max_length=32, null=True, default=lambda: secrets.token_urlsafe(nbytes=32)
+        # Since sharing keys are base64 encoded, we use the base64 length formula here
+        max_length=(((4 * settings.SHARING_KEY_LENGTH) // 3) + 3) & ~3,
```
Comment (diurnalist) [inline trovi/models.py:60]: & ~3? That is just black magic...

**Review by diurnalist — APPROVED** [review]
Nice, that ended up being quite clean. I like the various frameworks it provides to plug in to, like permissions.

File: docker/dev.Dockerfile (hunk 2 above)
Comment (super-cooper) [reply docker/dev.Dockerfile:1]: Good idea!

File: trovi/api/filters.py (hunk 9 above)
Comment (super-cooper) [reply trovi/api/filters.py:33]: Updated date, I assume? Since the created_date switch already exists?

File: trovi/api/serializers.py (hunk 10 above)
Comment (super-cooper) [reply trovi/api/serializers.py:90]: It informs the serializer that it is serializing multiple tags, and will therefor return a list.

**Review by super-cooper — COMMENTED** [review]
Merging now!
